InvokeArgumentValidator changes

Momo-Not-Emo opened the pull request:
1. Adapt InvokeArgumentValidator and add test cases. The original version primarily focuses on verifying the number of arguments. 
2. Implement argument type checking by including `rt.jar` in the analysis input locations, as the validator relies on typeHierarchy. Add a corresponding test case for this scenario.
3. Rectify a bug in `JavaIdentifierFactory.getBoxedType(...)` related to the int type. Replace the erroneous `java.lang.Int` with the correct type, which is `java.lang.Integer`.
4. Add support to check implicit conversion for primitive types.
5. Add a workaround check in `ArchiveBasedAnalysisInputLocation.getClassSource(...)` to avoid casting exception when passing an anonymous `ClassType` object which cannot be casted to `JavaClassType` directly.

Momo-Not-Emo committed:
InvokeArgumentValidator changes

swissiety commented on sootup.core/src/main/java/sootup/core/validation/InvokeArgumentValidator.java:
```diff
@@ -22,31 +22,42 @@
+            if (stmt.containsInvokeExpr()) {
+                AbstractInvokeExpr invExpr =
+                        stmt.getInvokeExpr();
+                MethodSignature callee = invExpr.getMethodSignature();
+                if (invExpr.getArgCount() !=
+                        callee.getParameterTypes().size()) {
```
This condition is always false as they point to the same data - we need to get a SootMethod via the view.

Momo-Not-Emo replied:
> This condition is always false as they point to the same data - we need to get a SootMethod via the view.

The condition is not always true. In the `testInvokeArgumentValidatorFailure` case, the Jimple statement `l1 = virtualinvoke l0.<InvokeArgumentValidator: java.lang.String join(int,java.lang.String)>(1)` is constructed. Here, `invExpr.getArgCount()` evaluates to 1 (indicating one argument passed) while `callee.getParameterTypes().size()` evaluates to 2 (indicating two arguments defined in the method signature). They point to different data.

swissiety replied:
I see - I was already thinking about the actual method definition and was confused that no retrieving of a method was done.

swissiety commented on sootup.core/src/main/java/sootup/core/validation/InvokeArgumentValidator.java:
```diff
@@ -22,31 +22,42 @@
+                        callee.getParameterTypes().size()) {
+                    validationException.add(new ValidationException(stmt, "Invalid number of arguments"));
+                }
+            }
```
An additional check for correct argument types (+inheritance - use the view.typehierarchy() ) would be nice, too.

swissiety commented on sootup.tests/src/test/java/sootup/tests/validator/InvokeArgumentValidatorTest.java:
```diff
@@ -0,0 +1,114 @@
+        classes = new HashSet<>(); // Set to track the classes to check
+
+        for (SootClass aClass : jimpleView.getClasses()) {
+            if (!aClass.isLibraryClass()) {
```
isLibraryClass check is not necessary as only an AnalysisInputLocation with SourceType.Application is added to the view

Momo-Not-Emo replied:
you are right. This code is copied from an existing validator test. I'll simplify it.

swissiety commented on sootup.tests/src/test/java/sootup/tests/validator/InvokeArgumentValidatorTest.java:
```diff
@@ -0,0 +1,114 @@
+    public void testInvokeArgumentValidatorSuccess() {
+        List<ValidationException> validationExceptions_success;
+
+        validationExceptions_success =
+                invokeArgumentValidator.validate(
+                        classes.stream()
+                                .filter(c -> c.getType().getClassName().equals("InvokeArgumentValidator"))
+                                .findFirst()
+                                .get()
+                                .getMethods()
+                                .stream()
+                                .filter(m -> m.getName().equals("invokeArgumentValidator_success"))
+                                .map(m -> m.getBody())
+                                .findFirst()
+                                .get(),
```
you can simplify accessing the methodBody: `view.getMethod( view.getIdentifierFactory().parseMethod("<InvokeArgumentValidator: void invokeArgumentValidator_success()>") ).get().getBody()`

Momo-Not-Emo committed:
check argument types and fix some bugs

1. Refactor the test case code as per the review comments for simplicity.
2. Implement argument type checking by including `rt.jar` in the analysis input locations, as the validator relies on typeHierarchy. Add a corresponding test case for this scenario.
3. Rectify a bug in `JavaIdentifierFactory.getBoxedType(...)` related to the int type. Replace the erroneous `java.lang.Int` with the correct type, which is `java.lang.Integer`.
4. Address a bug in `ArchiveBasedAnalysisInputLocation.getClassSource(...)` where attempting to cast `ClassType` to `JavaClassType` is incorrect. Update the code to handle `ClassType` appropriately, as it cannot be directly cast to `JavaClassType`.

Momo-Not-Emo committed:
Optimize classTypeFieldRefValidator existence check

Speed up the class search process by limiting the search scope within application classes, since classTypeFieldRefValidator does not exist in rt.jar

swissiety commented on sootup.core/src/main/java/sootup/core/validation/InvokeArgumentValidator.java:
```diff
@@ -22,31 +22,72 @@
+                    while (iterArgs.hasNext() && iterParameters.hasNext()) {
+                        ClassType argType = getClassType(identifierFactory, iterArgs.next().getType());
+                        ClassType paramType = getClassType(identifierFactory, iterParameters.next());
+                        if (argType != paramType && (!typeHierarchy.contains(paramType)
```
I am not sure the Typehierarchy can handle primitive type relations / BoxedType compatibility - maybe we have to do that seperately via [type promotion](https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html)

Momo-Not-Emo replied:
`Typehierarchy` doesn't consider "int" to be an instance of "Number", that's why I cast the primitive type to its boxed type in "getClassType(...)". 

Actually, in `Typehierarchy.isSubType(...)`, the comment says "subtyping applies to ReferenceTypes only". There is a `PrimitiveHierarchy` class, but its static method `isAncestor(...)` only checks the ancestor-relationship between two primitive types

swissiety replied:
But this does not handle cases like: `int` is used as an argument of a `double` parameter (while double -> int is not allowed implicitly) - we should not fail in these cases

Momo-Not-Emo replied:
> But this does not handle cases like: `int` is used as an argument of a `double` parameter (while double -> int is not allowed implicitly) - we should not fail in these cases

I have added implicit conversion checks

swissiety commented on sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/ArchiveBasedAnalysisInputLocation.java:
```diff
@@ -99,8 +99,13 @@ public Optional<JavaSootClassSource> getClassSource(@Nonnull ClassType type, @No
       FileSystem fs = fileSystemCache.get(path);
       final Path archiveRoot = fs.getPath("/");
+      JavaClassType javaClassType = null;
+      if (type instanceof JavaClassType)
+        javaClassType = (JavaClassType) type;
+      else
+        javaClassType = new JavaClassType(type.getClassName(), type.getPackageName());
       return getClassSourceInternal(
-          (JavaClassType) type, archiveRoot, new AsmJavaClassProvider(view));
+          javaClassType, archiveRoot, new AsmJavaClassProvider(view));
```
what happened here?

Momo-Not-Emo replied:
> what happened here?

It throws exception that an **anonymous** ClassType cannot be cast to JavaClassType:

```java
      ClassType classTypeFieldRefValidator =
                new ClassType() {
                    @Override
                    public boolean isBuiltInClass() {
                        return false;
                    }

                    @Override
                    public String getFullyQualifiedName() {
                        return "jimple.InvokeArgumentValidator";
                    }

                    @Override
                    public String getClassName() {
                        return "InvokeArgumentValidator";
                    }

                    @Override
                    public PackageName getPackageName() {
                        return new PackageName("jimple");
                    }
                };

       ....

        // rt.jar is required since the validator uses typeHierarchy
        DefaultRTJarAnalysisInputLocation defaultRTJarAnalysisInputLocation = new DefaultRTJarAnalysisInputLocation();
        jimpleView = new JimpleView(Arrays.asList(jimpleInputLocation, defaultRTJarAnalysisInputLocation));

        final Optional<SootClass> classSource1 = jimpleView.getClass(classTypeFieldRefValidator);
        assertFalse(classSource1.isPresent());
```

It will search for the `classTypeFieldRefValidator` within both application classes and library classes (rt.jar). Subsequently, it will prompt the casting exception.

In the SootUp class hierarchy, `ClassType` only has one specific type under it, which is `JavaClassType`. I've restricted the search to only look within the application classes. As a result, it won't send an anonymous  `ClassType` object to `ArchiveBasedAnalysisInputLocation` anymore. Should we keep this type check ?

swissiety replied:
We should address this issue in whole in another PR - please add a comment that this is a temporary hack

swissiety commented on sootup.java.core/src/main/java/sootup/java/core/JavaIdentifierFactory.java:
```diff
@@ -237,6 +237,8 @@ public Collection<PrimitiveType> getAllPrimitiveTypes() {
   @Nonnull
   public JavaClassType getBoxedType(@Nonnull PrimitiveType primitiveType) {
     String name = primitiveType.getName();
+    if (primitiveType instanceof PrimitiveType.IntType)
+      name = "integer";
```
lets add a getBoxedName() to the primitive types?

Momo-Not-Emo committed:
add a getBoxedName() to the primitive types

Momo-Not-Emo committed:
Check implicit conversion between primitive type

JonasKlauke commented on sootup.core/src/main/java/sootup/core/validation/InvokeArgumentValidator.java:
```diff
@@ -22,31 +22,80 @@
+                            continue;
+                        }
+                        // other cases
+                        ClassType argClassType = getClassType(identifierFactory, argType);
```
After the primitive type check the argument could still have the type: arraytype. You can not assume it is automatically a classtype.

JonasKlauke commented on sootup.tests/src/test/resources/validator/jimple/InvokeArgumentValidator.jimple:
```diff
@@ -0,0 +1,77 @@
+
+        return;
+    }
+
```
add a testcase for array types. Primitive type array, Classtype array and both cases also as multi-dimensional array

JonasKlauke commented on sootup.java.bytecode/src/main/java/sootup/java/bytecode/inputlocation/ArchiveBasedAnalysisInputLocation.java:
```diff
@@ -99,8 +99,15 @@ public Optional<JavaSootClassSource> getClassSource(@Nonnull ClassType type, @No
+      JavaClassType javaClassType = null;
+
+      // This is a temporary workaround to prevent a casting exception when passing an anonymous ClassType object
+      if (type instanceof JavaClassType)
```
What type do you get here in case of an anoynmous class?

Momo-Not-Emo replied:
> What type do you get here in case of an anoynmous class?
```
 ClassType classTypeFieldRefValidator =  new ClassType() {...}. // implement the abstract methods
```
It is a ClassType object but not an instance of JavaClassType. I observed that this code has been removed in the current test code.

JonasKlauke reviewed (changes requested):
Thanks for the help to improve SootUp 👍

swissiety committed:
streamline Validator, code style

swissiety committed:
Merge branch 'develop' into invk-arg-validator

swissiety committed:
adapt to upstream changes in develop

swissiety committed:
avoid classcastexceptions

swissiety committed:
simplify testcases

swissiety committed:
remove cce leftover

swissiety committed:
style

Momo-Not-Emo committed:
Merge branch 'develop' into invk-arg-validator
